test/test_images/sendevent/main.go

```diff
@@ -14,89 +14,119 @@ See the License for the specific language governing permissions and
 	"github.com/google/uuid"
+	"github.com/knative/pkg/cloudevents"
 )
 
+type Heartbeat struct {
+	Sequence int    `json:"id"`
+	Data     string `json:"data"`
+}
+
 var (
-	context cloudevents.EventContext
-	webhook string
-	data    string
+	sink      string
+	data      string
+	eventID   string
+	eventType string
+	source    string
+	periodStr string
+	delayStr  string
 )
 
 func init() {
-	flag.StringVar(&context.EventID, "event-id", "", "Event ID to use. Defaults to a generated UUID")
-	flag.StringVar(&context.EventType, "event-type", "google.events.action.demo", "The Event Type to use.")
-	flag.StringVar(&context.Source, "source", "", "Source URI to use. Defaults to the current machine's hostname")
-	flag.StringVar(&data, "data", `{"hello": "world!"}`, "Event data")
+	flag.StringVar(&sink, "sink", "", "The sink url for the message destination.")
+	flag.StringVar(&data, "data", "", "Special data.")
+	flag.StringVar(&eventID, "event-id", "", "Event ID to use. Defaults to a generated UUID")
+	flag.StringVar(&eventType, "event-type", "knative.eventing.test.e2e", "The Event Type to use.")
+	flag.StringVar(&source, "source", "", "Source URI to use. Defaults to the current machine's hostname")
+	flag.StringVar(&periodStr, "period", "5", "The number of seconds between messages.")
+	flag.StringVar(&delayStr, "delay", "5", "The number of seconds to wait before sending messages.")
+}
+
+func parseDurationStr(durationStr string, defaultDuration int) time.Duration {
+	var duration time.Duration
+	if d, err := strconv.Atoi(durationStr); err != nil {
+		duration = time.Duration(defaultDuration) * time.Second
+	} else {
+		duration = time.Duration(d) * time.Second
+	}
+	return duration
 }
 
 func main() {
 	flag.Parse()
+	period := parseDurationStr(periodStr, 5)
+	delay := parseDurationStr(delayStr, 5)
 
-	if len(flag.Args()) != 1 {
-		fmt.Println("Usage: sendevent [flags] <webhook>\nFor details about valid flags, run sendevent --help")
-		os.Exit(1)
+	if delay > 0 {
+		log.Printf("will sleep for %s", delay)
+		time.Sleep(delay)
+		log.Printf("awake, contining")
 	}
 
-	fmt.Println("Sleeping for 5 seconds")
-	//TODO maybe required for istio warmup?
-	time.Sleep(5 * time.Second)
-	fmt.Println("Done sleeping")
-
-	webhook := flag.Arg(0)
+	if eventID == "" {
+		eventID = uuid.New().String()
+	}
 
-	var untyped map[string]interface{}
-	if err := json.Unmarshal([]byte(data), &untyped); err != nil {
-		fmt.Println("Currently sendevent only supports JSON event data")
-		os.Exit(1)
+	if source == "" {
+		source = "localhost"
 	}
 
-	fillEventContext(&context)
-	req, err := cloudevents.NewRequest(webhook, untyped, context)
-	if err != nil {
-		fmt.Printf("Failed to create request: %s", err)
-		os.Exit(1)
+	hb := &Heartbeat{
+		Sequence: 0,
+		Data:     data,
 	}
-	fmt.Printf("requesting: %#v", req)
-	res, err := http.DefaultClient.Do(req)
-	if err != nil {
-		fmt.Printf("Failed to send event to %s: %s\n", webhook, err)
-		os.Exit(1)
+	ticker := time.NewTicker(period)
```

Comment:
Also uncertain about this change. Disadvantage of infinite retries in the sender is that it masks instability. A channel implementation that drops 90% of messages will pass this e2e test.

My original thought was that this should attempt to send a single event and exit with success or not based on that result. Then the pod restart strategy can take care of retrying, and we could maybe pass or fail the e2e test based on the number of restarts required.  But again, I don't feel strongly about this - I just want it to work.